Game manager idea

Hassan-A opened the pull request:
recreating everything with the idea of having a game manager script to simplify communication between objects. As well as for future work as we will need to for level data and patient types ext.. 
This merge will require conflicts to be taken completely from this branch or issues will occur.

Hassan-A committed:
Manually add bed

Hassan-A committed:
Add State class, Create object dictionary list

Hassan-A committed:
Merge 'main' into Game-Manager-Idea with working systems

Hassan-A committed:
Add Patient Queue, Patient Movement To Bed

Algruun commented on Hospital Saviour/Assets/Player(s)/Player1.cs:
```diff
@@ -6,18 +6,25 @@
 public class Player1 : MonoBehaviour
```
Does it means that we would have class `Player2`? Is it possible to utilize the same class for both players? maybe something like a factory
just question :)

Hassan-A replied:
We haven't reached that in the list, but i would assume we will refactor such that we don't need player2 class. A factory would be overkill for 2 players i  think.

Algruun commented on Hospital Saviour/Assets/Player(s)/Player1.cs:
```diff
@@ -35,29 +42,78 @@ public void OnMove(InputValue input)
+        {
+            if(itemType == "Folder")
+            {
+                foreach (GameObject go in collidingObjects)
+                {
+                    if (manager.objectList.Contains(go))
+                    {
```
This looks the same for both branches of if statement, maybe it can be moved to the top and content of this for loop will be selected with if/params?

Hassan-A replied:
not sure what you mean, can you give example

Algruun replied:
added suggestions to clarify what I meant, I hope it's more clear, haven't used github suggestions before

Algruun commented on Hospital Saviour/Assets/Player(s)/Player1.cs:
```diff
@@ -35,29 +42,78 @@ public void OnMove(InputValue input)
+                        Bed b = go.GetComponent<Bed>();
+                        if (b && b.isActive && b.isInteractable && !b.hasFolder && !b.isOccupied)
+                        {
+                            Folder f = item.GetComponent<Folder>();
+                            f.transferTo(go);
+                            f.changePosToBed();
+                            b.folderDropOff(item);
+                            f.patientOwner.GetComponent<Patient>().folderPlaced(go);
+                            manager.removeFromQueue(f.patientOwner);
+                            isCarrying = false;
+                            item = null;
+                            itemType = "";
```
What do you think about making this a function?

Algruun commented on Hospital Saviour/Assets/Player(s)/Player1.cs:
```diff
@@ -35,29 +42,78 @@ public void OnMove(InputValue input)
+                    {
+                        if (p.isHoldingFolder)
+                        {
+                            isCarrying = true;
+                            item = p.folder;
+                            itemType = "Folder";
+                            p.releaseFolder();
+                            Folder i = item.GetComponent<Folder>();
+                            i.transferTo(gameObject);
+                            i.changePosToPlayer();
```
same about function

Hassan-A replied:
sure, after merge we can do so :), i just don't want to add extra to the already complicated merge

Algruun commented on Hospital Saviour/Assets/Scripts/GameManager.cs:
```diff
@@ -0,0 +1,133 @@
+        //Instatiate Inactive beds
+        for (int i = 0; i < inActiveBedCount; i++)
+        {
+            GameObject newBed = Instantiate(bedPrefab, inActiveBedParent, false);
+            newBed.transform.position += new Vector3(0, 0, -bedSeperation * i);
+            var bedRenderers = newBed.GetComponentsInChildren<MeshRenderer>();
+            foreach (MeshRenderer renderer in bedRenderers)
+            {
+                renderer.material = inActiveMat;
+            }
+            //Place bed into object list
+            objectList.Add(newBed);
```
Same question about making it a separate function :)
This might help avoid having issues later

Algruun commented on Hospital Saviour/Assets/Scripts/GameManager.cs:
```diff
@@ -0,0 +1,133 @@
+        //Instatiate Active beds
+        for (int i = 0; i < activeBedCount; i++)
+        {
+            float bedStartingPos = inActiveBedCount * -bedSeperation;
+            GameObject newBed = Instantiate(bedPrefab, activeBedParent, false);
+            newBed.transform.position += new Vector3(0, 0, bedStartingPos - (bedSeperation * i));
+            //Place bed into object list
+            objectList.Add(newBed);
+            Bed b = newBed.GetComponent<Bed>();
+            b.isActive = true;
+            b.isInteractable = true;
```
Probably might be useful to merge it with `Instatiate Inactive beds` into a function with param active = true/false. Since there should be big interlap in logic between instatiating inactive and active beds

Algruun commented on Hospital Saviour/Assets/Scripts/BaseInteractable.cs:
```diff
@@ -1,18 +0,0 @@
-﻿using System.Collections;
```
What's the benefit of removing this class?
Looks like bed and patient are still interactable and it might be useful to have a base class for them with some basic logic. Am I missing something?

Hassan-A replied:
the point of this branch was to remove composition classes and try and use a game manager class for simplicity. therefore this was not needed anymore and will not be used in this branch.

Algruun reviewed (commented):
Just asking questions :)

Hassan-A committed:
Add Icon system and first type

Algruun commented on Hospital Saviour/Assets/Player(s)/Player1.cs:
```diff
@@ -35,29 +42,78 @@ public void OnMove(InputValue input)
 
     public void OnInteract()
     {
-        if(interactable != null)
+        if (isCarrying)
+        {
+            if(itemType == "Folder")
+            {
+                foreach (GameObject go in collidingObjects)
+                {
+                    if (manager.objectList.Contains(go))
+                    {
+                        Bed b = go.GetComponent<Bed>();
+                        if (b && b.isActive && b.isInteractable && !b.hasFolder && !b.isOccupied)
+                        {
+                            Folder f = item.GetComponent<Folder>();
+                            f.transferTo(go);
+                            f.changePosToBed();
+                            b.folderDropOff(item);
+                            f.patientOwner.GetComponent<Patient>().folderPlaced(go);
+                            manager.removeFromQueue(f.patientOwner);
+                            isCarrying = false;
+                            item = null;
+                            itemType = "";
+
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+        else
```
```suggestion
       foreach (GameObject go in collidingObjects)
        {
            if (manager.objectList.Contains(go))
            {
                if (isCarrying)
                {
                    if (itemType == "Folder")
                    {
                        Bed b = go.GetComponent<Bed>();
                        if (b && b.isActive && b.isInteractable && !b.hasFolder && !b.isOccupied)
                        {
                            Folder f = item.GetComponent<Folder>();
                            f.transferTo(go);
                            f.changePosToBed();
                            b.folderDropOff(item);
                            f.patientOwner.GetComponent<Patient>().folderPlaced(go);
                            manager.removeFromQueue(f.patientOwner);
                            isCarrying = false;
                            item = null;
                            itemType = "";

                            break;
                        }
                    }
                    else
                    {
                        // Optional?
                        break;
                    }
                }
                else
```

Algruun commented on Hospital Saviour/Assets/Player(s)/Player1.cs:
```diff
@@ -35,29 +42,78 @@ public void OnMove(InputValue input)
+        }
+        else
         {
-            interactable.GetComponent<BaseInteractable>().MainInteract(gameObject);
+            foreach (GameObject go in collidingObjects)
+            {
+                if (manager.objectList.Contains(go))
+                {
+                    Patient p = go.GetComponent<Patient>();
+                    print(p);
+                    if (p && p.isInteractable)
+                    {
+                        if (p.isHoldingFolder)
+                        {
+                            isCarrying = true;
+                            item = p.folder;
+                            itemType = "Folder";
+                            p.releaseFolder();
+                            Folder i = item.GetComponent<Folder>();
+                            i.transferTo(gameObject);
+                            i.changePosToPlayer();
+
+                            break;
+                        }
+                    }
+                }
+            }
```
```suggestion
                    {
                    Patient p = go.GetComponent<Patient>();
                    print(p);
                    if (p && p.isInteractable)
                    {
                        if (p.isHoldingFolder)
                        {
                            isCarrying = true;
                            item = p.folder;
                            itemType = "Folder";
                            p.releaseFolder();
                            Folder i = item.GetComponent<Folder>();
                            i.transferTo(gameObject);
                            i.changePosToPlayer();

                            break;
                        }
                    }
                }
            }
```

Hassan-A replied:
looks fine, again we can definitely do this but after the merge because it is currently stable and we would need to test the other way.

mecoat committed:
Note location moved, head1 type added, countdown speed added

Snoop14 committed:
Added Soup

Added soup Machine, soup, and interaction with soup machine

Hassan-A committed:
Icon fix

Hassan-A committed:
Fix Folder Bed Placement

Hassan-A committed:
Merge branch 'main' into Game-Manager-Idea

Hassan-A committed:
resolving conflicts for merge into main

Hassan-A committed:
Icon clarity

Algruun committed:
Add Patient to get into bed

Algruun committed:
Merge branch 'Game-Manager-Idea' of https://github.com/Snoop14/Hospital-Saviour-Game into Game-Manager-Idea

Hassan-A committed:
Add bone type

Algruun commented:
Can we merge this branch? It's become hard to work on separate features :(

Snoop14 committed:
Soup transfer added

Soup can now be transferred to a patient on bed(through bed) also fixed patient not getting in bed
